### PR TITLE
Expressions should not appear twice in the ast

### PR DESCRIPTION
**Pull request by kripken**
This adds validation for the tree structure in our IR, that each node must have exactly one parent, and so each node must appear exactly once and no more. Turns out we had some errors in that regard, one actual bug and one test bug, both fixed. Also added some docs to the readme.

### CODE REVIEW

File: src/wasm/wasm-validator.cpp [hunk 1]
```diff
@@ -559,6 +559,26 @@ void WasmValidator::visitFunction(Function *curr) {
...
   labelNames.clear();
+  // expressions must not be seen more than once
+  struct Walker : public PostWalker<Walker, UnifiedExpressionVisitor<Walker>> {
+    std::unordered_set<Expression*>& seen;
+    std::vector<Expression*> dupes;
+
+    Walker(std::unordered_set<Expression*>& seen) : seen(seen) {}
+
+    void visitExpression(Expression* curr) {
+      if (seen.find(curr) == seen.end()) {
```
Comment (dschuff) [inline src/wasm/wasm-validator.cpp]: Can you just do
`if (seen.insert(curr).second) dupes.push_back(curr);`

Comment (kripken) [reply src/wasm/wasm-validator.cpp]: I suppose, but I find the `.second` notation hard to read myself (a shame it's not `.inserted`). Is it just me?

Comment (dschuff) [reply src/wasm/wasm-validator.cpp]: No, I agree the API is ugly.
Could always do something like
```
bool inserted;
std::tie<std::ignore, inserted> = seen.insert(curr);
if (!inserted) seen.insert(curr);
```

Comment (kripken) [reply src/wasm/wasm-validator.cpp]: Thanks, I like that way, updated.

**Pull request merged**